lib/utils/fanoutbuffer/buffer.go

```diff
@@ -0,0 +1,426 @@
+	// of that element.
+	pos uint64
+
+	// len is the total number of elements in the ring (i.e. the index of the newest element is at `(pos + len - 1) % Capacity`).
+	len uint64
+
+	// ring is the fixed-size circular buffer that serves as our primary storage location.
+	ring []entry[T]
+
+	// overflowStart is the time at which we began needing to overflow items.
+	overflowStart time.Time
+
+	// overflow is a variable length buffer of items that still need to be observed.
+	overflow []entry[T]
+}
+
+// entry represents an item in a buffer
+type entry[T any] struct {
+	item T
+	// wait is the number of cursors that still need to observe this item. each cursor
+	// decrements this value when it observes the item. items with wait==0 are cleaned
+	// up. Note that this atomic is only safe to decrement while read lock is held. Entries
+	// may be moved while write lock is held.
+	wait atomic.Uint64
+}
+
+// NewBuffer creates a new fanout buffer instance.
+func NewBuffer[T any](cfg Config) *Buffer[T] {
+	cfg.SetDefaults()
+	return &Buffer[T]{
+		cfg:    cfg,
+		notify: make(chan struct{}),
+		ring:   make([]entry[T], cfg.Capacity),
+	}
+}
+
+// Close permanently closes the fanout buffer. Note that all cursors are terminated immediately
+// and may therefore miss items that had not been read at the time Close is called.
+func (b *Buffer[T]) Close() {
+	b.write(func() {
+		b.closed = true
+		close(b.notify)
+	})
+}
+
+// NewCursor gets a new cursor into the buffer. Cursors *must* be closed as soon as they are
+// no longer being read from. Stale cursors may cause performance degredation.
+func (b *Buffer[T]) NewCursor() *Cursor[T] {
+	var c uint64
+	b.write(func() {
+		if b.closed {
+			return
+		}
+		c = b.pos + b.len
+		b.cursors++
+	})
+
+	cursor := &Cursor[T]{
+		buf:    b,
+		cursor: c,
+	}
+	runtime.SetFinalizer(cursor, finalizeCursor[T])
+	return cursor
+}
+
+// Append appends to the buffer and wakes all dormant cursors.
+func (b *Buffer[T]) Append(items ...T) {
+	var notify chan struct{}
+
+	b.write(func() {
+		if b.closed {
+			panic("Append called on closed fanout buffer")
+		}
+
+		// free up slots that are no longer needed
+		b.cleanupSlots()
+
+		for _, item := range items {
+			// ensure that there is at least one free slot in ring
+			b.ensureSlot()
+
+			// insert item into next free slot in ring
+			idx := int((b.pos + b.len) % b.cfg.Capacity)
+			b.ring[idx].item = item
+			b.ring[idx].wait.Store(b.cursors)
+			b.len++
+		}
+
+		notify = b.notify
+		b.notify = make(chan struct{})
+	})
+
+	// notify all waiting cursors. we do this outside of the
+	// lock to (theoretically) reduce potential contention.
+	close(notify)
+}
+
+// cleanupSlots frees up enries that have been seen by all cursors and clears the backlog if
+// it has exceeded its grace period.
+func (b *Buffer[T]) cleanupSlots() {
+	// reset overflow if it is past its grace period
+	if !b.overflowStart.IsZero() && b.cfg.Clock.Now().After(b.overflowStart.Add(b.cfg.GracePeriod)) {
```

Comment:
How does this work if two cursors are causing overflow to different degrees?

Let's say once cursor is occasionally dipping into overflow due to intermittent connection issues, but the cursor is always read within 5 seconds. If this is the only cursor, the buffer will work as expected since the overflow will naturally be cleared and reset without the intervention of the grace period logic.

However, if a second cursor does fail to read from overflow within the grace period, the overflow will reset.

If the first cursor happened to be reading from overflow at this time, won't it also be marked as unhealthy, even though it is actually keeping up well enough? Or do we just not care enough about cursors in overflow, instead prioritizing efficiency?

It seems like we should accommodate intermittently overflowing cursors by marking each entry in the overflow with it's overflow time, and only cutting one entry from the overflow at a time when they exceed the grace period.

```suggestion
	// trim expired items from overflow
	var clearOverflowTo int
	for i := 0; i < len(b.overflow); i++
		clearOverflowTo = i
		if !time.Now().After(entry.overflowExpiry) {
			// oldest overflow entry is not expired, no need to iterate further.
			break
		}
	}
	b.overflow = append(b.overflow[:0], b.overflow[clearOverflowTo:]...)
```